File: articles/diagnostics-residuals-and-model-error/fortran/diagnostic_summary.f90
program diagnostic_summary
  implicit none

  integer, parameter :: n = 10
  integer :: i, disagreements
  real(8) :: observed(n), predicted(n), threshold(n)
  real(8) :: residual, sum_abs, sum_sq, bias, max_abs, rmse, mae

  observed = (/82.0d0,79.5d0,77.0d0,74.3d0,71.5d0,69.2d0,67.8d0,65.5d0,63.0d0,61.1d0/)
  predicted = (/81.5d0,80.2d0,78.4d0,75.6d0,72.8d0,71.0d0,69.8d0,68.0d0,66.4d0,65.2d0/)
  threshold = (/70.0d0,70.0d0,70.0d0,70.0d0,70.0d0,70.0d0,70.0d0,70.0d0,70.0d0,70.0d0/)

  sum_abs = 0.0d0
  sum_sq = 0.0d0
  bias = 0.0d0
  max_abs = 0.0d0
  disagreements = 0

  do i = 1, n
    residual = observed(i) - predicted(i)
    sum_abs = sum_abs + abs(residual)
    sum_sq = sum_sq + residual * residual
    bias = bias + residual
    max_abs = max(max_abs, abs(residual))
    if ((observed(i) < threshold(i) .and. predicted(i) >= threshold(i)) .or. &
        (observed(i) >= threshold(i) .and. predicted(i) < threshold(i))) then
      disagreements = disagreements + 1
    end if
  end do

  rmse = sqrt(sum_sq / real(n, 8))
  mae = sum_abs / real(n, 8)
  bias = bias / real(n, 8)

  print '(A)', 'mean_error mae rmse max_abs_error decision_disagreements'
  print '(F8.4,1X,F8.4,1X,F8.4,1X,F8.4,1X,I4)', bias, mae, rmse, max_abs, disagreements

end program diagnostic_summary

File: articles/diagnostics-residuals-and-model-error/c/diagnostic_summary.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int time;
    const char *group;
    double observed;
    double predicted;
    double threshold;
} Observation;

int main(void) {
    Observation data[] = {
        {1, "baseline", 82.0, 81.5, 70.0},
        {2, "baseline", 79.5, 80.2, 70.0},
        {3, "baseline", 77.0, 78.4, 70.0},
        {4, "baseline", 74.3, 75.6, 70.0},
        {5, "threshold", 71.5, 72.8, 70.0},
        {6, "threshold", 69.2, 71.0, 70.0},
        {7, "threshold", 67.8, 69.8, 70.0},
        {8, "stress", 65.5, 68.0, 70.0},
        {9, "stress", 63.0, 66.4, 70.0},
        {10, "stress", 61.1, 65.2, 70.0}
    };

    const int n = 10;
    double sum_abs = 0.0;
    double sum_sq = 0.0;
    double bias = 0.0;
    double max_abs = 0.0;
    int disagreements = 0;

    for (int i = 0; i < n; ++i) {
        double residual = data[i].observed - data[i].predicted;
        double abs_error = fabs(residual);
        bool observed_below = data[i].observed < data[i].threshold;
        bool predicted_below = data[i].predicted < data[i].threshold;

        sum_abs += abs_error;
        sum_sq += residual * residual;
        bias += residual;

        if (abs_error > max_abs) {
            max_abs = abs_error;
        }

        if (observed_below != predicted_below) {
            disagreements += 1;
        }
    }

    printf("mean_error,mae,rmse,max_abs_error,decision_disagreements\n");
    printf("%.6f,%.6f,%.6f,%.6f,%d\n",
           bias / (double)n,
           sum_abs / (double)n,
           sqrt(sum_sq / (double)n),
           max_abs,
           disagreements);

    fprintf(stderr, "c diagnostic_summary complete\n");
    return EXIT_SUCCESS;
}
